Ad7192 out of staging

mirceacaprioru opened the pull request:
Modifications in order to get the driver out of staging.

stefpopa commented:
As a general remark, you will also need a DT binding under Documentation/devicetree/bindings/iio/adc

mirceacaprioru commented:
I will add that to but not sure where to put it Documentation/devicetree/bindings/staging/iio or Documentation/devicetree/bindings/iio/adc

stefpopa commented:
> I will add that to but not sure where to put it Documentation/devicetree/bindings/staging/iio or Documentation/devicetree/bindings/iio/adc

 **Documentation/devicetree/bindings/iio/adc**
Add it in a separate commit in the same patch set where you move the driver out of staging.

stefpopa commented on drivers/staging/iio/adc/ad7192.c:
```diff
@@ -622,63 +643,38 @@ static const struct iio_chan_spec ad7193_channels[] = {
+	st->chop_en = of_property_read_bool(np, "adi,chop-enable");
+	st->buf_en = of_property_read_bool(np, "adi,buffer-enable");
+	st->unipolar_en = of_property_read_bool(np, "adi,unipolar-enable");
+	st->burnout_curr_en = of_property_read_bool(np, "adi,burnout-currents-enable");
```
Since all this properties are required, you also need to return an error in case they are missing.

mirceacaprioru replied:
These are all optional. No return needed.

stefpopa replied:
Ok. However, some of these properties do not belong in the dt binding. 
* refin2-pins-enable: you can add a second regulator in the device tree. If it exists, then you can enable refin2.
* rejection-60-Hz-enable and sinc3-filter-enable. You should use the IIO_CHAN_INFO_LOW_PASS_FILTER_3DB_FREQUENCY attribute to set the bw.
* buffer-enable: we can use the iio_triggered_buffer_setup().
* adi,unipolar-enable: it's probably fine, but soon a generic binding called `bipolar `will be added under Documentation/devicetree/bindings/iio/adc/adc.txt. You can follow this discussion: https://lore.kernel.org/patchwork/patch/1011859/.
* about adi,chop-enable and adi,burnout-currents-enable I am not sure, but you can try to check other bindings and see if there is anything similar.

stefpopa replied:
Some of the above remarks haven't been addressed.

mirceacaprioru replied:
are you referring to the unipolar option ?

stefpopa replied:
Yes, unipolar should be replaced with bipolar. Also, I think that you do no need `refin2-pins-enable`, `rejection-60-Hz-enable`  and` sinc3-filter-enable` anymore, since you found a better approach. Have you checked the other properties? To me they look like configuration options that should be left to user space.

stefpopa commented:
Another general remark: I think that we should review the whole driver before trying to send it upstream.

stefpopa commented on drivers/staging/iio/adc/ad7192.c:
```diff
@@ -701,11 +697,21 @@ static int ad7192_probe(struct spi_device *spi)
-	if (pdata->vref_mv)
-		st->int_vref_mv = pdata->vref_mv;
-	else if (voltage_uv)
+	if (voltage_uv)
```
Just curious: are the regulators optional? If yes, I think that we should use `devm_regulator_get_optional(`).

mirceacaprioru replied:
they are not optional.

stefpopa replied:
Therefore, should we continue if there is an error or no regulator in the dt?

commodo commented on drivers/staging/iio/adc/ad7192.c:
```diff
@@ -157,14 +156,23 @@
 	struct regulator		*dvdd;
+	struct clk			*mclk;
 	u16				int_vref_mv;
-	u32				mclk;
```
this patch does quite a lot in one go ;
i think splitting it up would be a good idea ;

first patch could do with just the clock changes ;

commodo commented on drivers/staging/iio/adc/ad7192.c:
```diff
@@ -157,14 +156,23 @@
 	u8				gpocon;
 	u8				devid;
+	u8				clock_sel;
+	bool				refin2_en;
```
these bool fields could be in a separate change that only moves them from the pdata struct to the state struct

commodo commented on drivers/staging/iio/adc/ad7192.c:
```diff
@@ -251,19 +259,31 @@ static int ad7192_setup(struct ad7192_state *st,
 	case AD7192_CLK_EXT_MCLK2:
-		if (ad7192_valid_external_frequency(pdata->ext_clk_hz)) {
-			st->mclk = pdata->ext_clk_hz;
+		ret = clk_prepare_enable(st->mclk);
```
this block of code could be moved into it's own function;
to reduce the code in this case block

mirceacaprioru commented:
Changelog V1->V2
- divided into multiple commits to be able to follow
- removed unnecessary dt bindings
- added 3db low pass filter 
- added regulator for vref

stefpopa commented on drivers/staging/iio/adc/ad7192.c:
```diff
@@ -405,6 +399,32 @@ static ssize_t ad7192_set(struct device *dev,
+	struct ad7192_state *st = iio_priv(indio_dev);
+	size_t len = 0;
+	int freq_avail[3];
+	int i;
```
nitpick: you can fit these variables on the same line

stefpopa commented on drivers/staging/iio/adc/ad7192.c:
```diff
@@ -405,6 +399,32 @@ static ssize_t ad7192_set(struct device *dev,
+	freq_avail[1] = mult_frac(st->fclk, 240, 1000);
+	freq_avail[2] = mult_frac(st->fclk, 272, 1000);
+
+	for (i = 0; i < 3; i++)
```
use ARRAY_SIZE(freq_avail)

stefpopa commented on drivers/staging/iio/adc/ad7192.c:
```diff
@@ -441,6 +463,59 @@ static unsigned int ad7192_get_temp_scale(bool unipolar)
+	}
+
+	ret = ad_sd_write_reg(&st->sd, AD7192_REG_MODE, 3, st->mode);
+		if (ret < 0)
```
nitpick: indentation looks wrong

stefpopa commented on arch/arm/boot/dts/overlays/rpi-ad7192-overlay.dts:
```diff
@@ -0,0 +1,76 @@
+				adi,buffer-enable;
+				adi,burnout-currents-enable;
+				adi,sinc3-filter-enable;
+				adi,unipolar-enable;
```
you can backport this [commit](https://git.kernel.org/pub/scm/linux/kernel/git/jic23/iio.git/commit/?h=testing&id=90c430d32c5fa08363b3db402df1b937901dd5b3) and use the `bipolar `property.

stefpopa commented on arch/arm/boot/dts/overlays/rpi-ad7192-overlay.dts:
```diff
@@ -0,0 +1,76 @@
+				clocks = <&ad7190_clk>;
+				clock-names = "clk";
+
+				adi,refin2-pins-enable;
```
I think you do not need this anymore

stefpopa commented:
> > I will add that to but not sure where to put it Documentation/devicetree/bindings/staging/iio or Documentation/devicetree/bindings/iio/adc
> 
> **Documentation/devicetree/bindings/iio/adc**
> Add it in a separate commit in the same patch set where you move the driver out of staging.
Bindings are still missing

stefpopa commented on drivers/staging/iio/adc/ad7192.c:
```diff
@@ -167,10 +167,8 @@ struct ad7192_state {
 	u8				devid;
```
Maybe you should add more information in your commit message when stating that
>These modifications improve overall performance

 You can also indicate the datasheet page where this information can be found.

stefpopa commented on drivers/staging/iio/adc/ad7192.c:
```diff
@@ -441,6 +463,59 @@ static unsigned int ad7192_get_temp_scale(bool unipolar)
+
+	switch (i) {
+	case 0:
+		st->f_order = 4;
```
Nitpick: I would use a define such as `#define AD7192_SINC4_FILTER 4`, but feel free to ignore this comment

stefpopa commented on drivers/staging/iio/adc/ad7192.c:
```diff
@@ -441,6 +463,59 @@ static unsigned int ad7192_get_temp_scale(bool unipolar)
+static int ad7192_set_3db_filter_freq(struct ad7192_state *st, int freq)
+{
+	int freq_avail[3];
+	int i, ret;
```
Nitpick: you can fit these on a single line

stefpopa commented on drivers/staging/iio/adc/ad7192.c:
```diff
@@ -441,6 +463,59 @@ static unsigned int ad7192_get_temp_scale(bool unipolar)
+		if (freq_avail[i] >= freq)
+			break;
+	if (i == ARRAY_SIZE(freq_avail))
+		return -EINVAL;
```
Another approach can be to allow the user to introduce whatever value he desires for the frequency and we can just find the closest match. Also, if the input value exceeds the upper limit, we can just set the max frequency.
